### ISSUE TITLE
BigWig Performance Regression

### ISSUE THREAD

**Issue by kcibul** [igvteam/igv#1505]
Hi IGV Team -- we've recently upgrade to IGV 2.17.x  from a commit of 2.16 that enables SSO (#1334).  After upgrading we found a significant performance degradation with BigWig files.  We load ~50 BigWigs from S3 and navigating to a new locus went from ~2 seconds to ~8 seconds.  The degradation is still present on a build off `master` as well.

I spent some time doing a few local builds of IGV between those two commits and the last performant build was:

[300ab06](https://github.com/igvteam/igv/commit/300ab06a3405ae4c61af69942c7bb06db1a89337)

the next commit 

[bcdd371](https://github.com/igvteam/igv/commit/bcdd3713cd1d530e50840860b4886bfe45118472)

appears to be a massive refactoring/rewrite of the BigWig support, so it looks suspicious but I am unable to run that commit as it fails with errors related to genome loading.  However given our problem is with BigWigs and the huge commit... my money is on something in this commit.

```
SEVERE [Mar 18,2024 17:16] [MessageUtils] java.lang.reflect.InaccessibleObjectException: Unable to make public org.broad.igv.feature.genome.load.GenomeConfig() accessible: module org.igv does not "exports org.broad.igv.feature.genome.load" to module com.google.gson
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Constructor.checkCanSetAccessible(Constructor.java:188)
	at java.base/java.lang.reflect.Constructor.setAccessible(Constructor.java:181)
	at com.google.gson@2.8.9/com.google.gson.internal.reflect.UnsafeReflectionAccessor.makeAccessible(UnsafeReflectionAccessor.java:44)
	at com.google.gson@2.8.9/com.google.gson.internal.ConstructorConstructor.newDefaultConstructor(ConstructorConstructor.java:103)
	at com.google.gson@2.8.9/com.google.gson.internal.ConstructorConstructor.get(ConstructorConstructor.java:85)
	at com.google.gson@2.8.9/com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:101)
	at com.google.gson@2.8.9/com.google.gson.Gson.getAdapter(Gson.java:489)
	at com.google.gson@2.8.9/com.google.gson.Gson.fromJson(Gson.java:962)
	at com.google.gson@2.8.9/com.google.gson.Gson.fromJson(Gson.java:928)
	at com.google.gson@2.8.9/com.google.gson.Gson.fromJson(Gson.java:877)
	at com.google.gson@2.8.9/com.google.gson.Gson.fromJson(Gson.java:848)
	at org.igv/org.broad.igv.feature.genome.load.JsonGenomeLoader.loadGenome(JsonGenomeLoader.java:54)
	at org.igv/org.broad.igv.feature.genome.GenomeManager.loadGenome(GenomeManager.java:198)
	at org.igv/org.broad.igv.feature.genome.GenomeManager.loadGenomeById(GenomeManager.java:176)
	at org.igv/org.broad.igv.ui.IGV$StartupRunnable.run(IGV.java:1930)
	at org.igv/org.broad.igv.util.LongRunningTask.call(LongRunningTask.java:72)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by jrobinso** [comment]
I'm impressed you can load 50 bigwigs from S3 with IGV.   That's not really a use case we design for or test.   If you can capture the network requests there might be a clue there, its entirely possible that optimizations for other use cases negatively impacts this one.   

It might be some time before I can look at this,  in the meantime I would recommend sticking with 2.16 or the last performant build.

**Comment by kcibul** [comment]
Thanks for the quick response -- we can take a look into that.  Was the in-sourcing of BigWig (and BBFile support, etc) just rearranging code or was it a brand new library/version. 

We're looking at ATAC, Chip-Seq style data and I'm curious if there's another format that would work better with IGV at that kind of scale?  Some folks have suggested making TDFs?

**Comment by jrobinso** [comment]
I'm not sure what you mean by "in sourcing", its always been internal code.   Many changes were required to efficiently support UCSC track hubs, this was the main driver.   There could be optimizations there that have negatively impacted the case of 50 bigwigs.   For example if more smaller requests are made instead of a few large requests this might not be noticeable for 5 bigwigs, but really noticeable for 50,  due to the overhead if each request.   But that is a wild guess.   The opposite could equally be true.     Its going to be some weeks before I can look at this. 

I recommend sticking with bigwig as its supported for more purposes than just IGV.   However TDF might be faster for IGV usage alone.

**Comment by kcibul** [comment]
Ah sorry -- that was a big commit!  I thought that `org.broad.igv.ucsc.bb` was all imported code from external library, but now I see it was just refactored/moved from `org.broad.igv.bb`.  Github is can be pretty bad showing moves as delete + add.

Anecdotally, it does appear to be something about parallelism as in the older IGV tracks seem to all appear in parallel at once but in 2.17.x they seem to appear serially one at a time.  

Any info we can gather I'll pass along in this ticket.  This is an important datatype for us, so if you have intuition about loading via S3 vs http (any reason to think one would be more performant) and BigWig vs TDF is great.  We'll still have BigWigs, but having another format that IGV would be worth it if it helps.

**Comment by jrobinso** [comment]
Hmm, the parallelism comment is interesting. That's a useful observation.

**Comment by maximilianh** [comment]
I'm not sure if this is relevant, but at UCSC we're loading bigWigs with
100 parallel threads. That has caused not too many problems so far and was
quite a bit faster.

**Comment by kcibul** [comment]
Hi @maximilianh -- Thanks, this is really helpful to know!  When you say "with 100 parallel threads" is that something you're specifying, or do you mean you load 100 samples at once?  Are you loading from S3, local filesystem or something else?

**Comment by maximilianh** [comment]
sorry I meant: The UCSC Genome browser uses 100 parallel threads, from
whatever webserver.

**Comment by jrobinso** [comment]
Yes I think I must have accidentally serialized something that happened in parallel before, based on @kcibul 's description. I'm sure this will be found, and its probably a single line of code somewhere.   I don't think we know that this problem is exclusive to bigwigs, btw, @kcibul do you have other file types  (that is other than bigwig / bigbed) that are loaded in these numbers?

**Comment by jrobinso** [comment]
Also, @kcibul , is this public data?  Having a session to reproduce the problem would save me some time.

**Comment by jrobinso** [comment]
@kcibul Actually I am seeing the opposite behavior on files loaded from https://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/.   The attached session, consisting of 100 bigwig files,  loads in about 6 seconds in version 2.17.3 on my rather slow internet.  It takes 80 seconds in version 2.16.2.    

[100_bigwigs.xml.zip](https://github.com/igvteam/igv/files/14653061/100_bigwigs.xml.zip)

**Comment by kcibul** [comment]
Interesting... so here's what I found.  I modified your XML so that it starts with `locus="chr11:62671495-62700195"` to begin with and then I navigate to "chr12:82,141,669-82,191,352"

2.17.3 takes ~100 seconds to load initially. The names of the tracks appear right away, and then the data for each track appears in batches.  Then takes another 90 seconds to navigate to the new locus with the tracks appearing sequentially again.  

2.16.2 about the same amount of time to load initially ~100 seconds, but the display empty (no track names) and then the tracks appear at once with the data.  Then when I navigate it takes ~5 seconds for the data to appear.

One theory -- Is it possible that 2.16.2 is downloading more of the data initially and then caching it and 2.17.3 is not?  Where we really noticed the slow down was in navigating around after loading.  Even just moving forwards/backwards along genes was remarkably different.  Ctrl-F on the gene track takes a few seconds per move in 2.16 but that full reload feeing in 2.17

**Comment by jrobinso** [comment]
`Is it possible that 2.16.2 is downloading more of the data initially`

Yes that's entirely possible.

**Comment by jrobinso** [comment]
I'm on this with JProfiler,  it will be fixed.   We seem to have lost some needed buffering for this use case, resulting in thousands of small http requests.   I don't know how easy the solution will be but there will be one.

**Comment by jrobinso** [comment]
@kcibul Could you give the snapshot build a try?  I found the problem, or a problem, the R Tree index was being read every time data was loaded.   Not really noticeable with a few big wigs, its fast,  but for 100 its not so fast.  You can download the snapshot here: https://igv.org/doc/desktop/#DownloadSnapshot/

**Comment by kcibul** [comment]
@jrobinso Seems to be much zippier in my hands, thanks for the fix!  We'll get it into the hands of more users and report back.

**Comment by kcibul** [comment]
@jrobinso -- feedback is all positive.  Is there a way to get a Mac package built with embedded JVM?  For some of our users, having a local mac java install is a barrier to entry.  Is it something I can do myself if there are no plans for a formal release on the horizon?

**Comment by helgathorv** [comment]
Hi @kcibul - we just released 2.17.4 that has the fix.
Download from https://igv.org/doc/desktop/#DownloadPage/
If it doesn't show 2.17.4, the browser cache may need clearing.

**Comment by kcibul** [comment]
Y'all are the best -- ❤️ !
